### PR TITLE
Improve computation of failed scores to not require increasing "miss" count

### PR DESCRIPTION
**Pull request by frenzibyte**
- [x] Depends on #19379 

In order to display failed scores with a sane total score and an accuracy that matches the one at the point of fail, `ScoringValues` is now included in local/online score models (as proposed by @smoogipoo).

To explain how this fixes the issues seen with failed scores:
 - https://github.com/ppy/osu/commit/583dbb0168eee9f64c4297ab30d3a7bdfbf72059: `ScoreProcessor` will use the score's scoring values and the maximum achievable scoring values to determine the total score in the `ComputeScore` methods.
 - https://github.com/ppy/osu/commit/071b94f8fa1040b7984c9a4966a5bf9c14e27b30: `ScoreManager.GetMaximumAchievableComboAsync` will use the score's maximum achievable scoring values to determine the max combo, rather than counting the statistics.
 - https://github.com/ppy/osu/commit/0cb0b4beb0b5f09e2e0e6eef9094a03e7dbcd03d: `ScoreInfo.GetStatisticsForDisplay` will use the score's maximum achievable scoring values to determine the total small/large ticks in the beatmap, rather than counting the statistics.

I'm not 101% sure about the inclusion of `{Large/Small}TicksCount` in `ScoringValues`, but it's required to display them properly in results screen. Open for suggestions, but I found this to be the least worst way to achieve it.

| failed | perfect |
|--------|---------|
|  | 

@smoogipoo Requesting your review in case I missed something here.

### CODE REVIEW

**Review by smoogipoo — CHANGES_REQUESTED** [review]
Aside from nits, I think this is probably okay.

Reason I haven't merged this yet is I've been meaning to integrate into osu-queue-score-statistics importer for osu-stable scores, but haven't gotten around to it.

I'm not sure about the tick stuff. Also not sure how that's going to work with stable yet. Because the goal at the end of the day is for these to never be null inside `ScoreInfo`/`SoloScoreInfo`.

File: osu.Game/Scoring/ScoringValues.cs [hunk 1]
```diff
@@ -20,24 +21,45 @@ public struct ScoringValues
...
 
         /// <summary>
-        /// The count of "basic" <see cref="HitObject"/>s. See: <see cref="HitResultExtensions.IsBasic"/>.
+        /// The count of <see cref="HitObject"/>s awarding "basic" hit results.
         /// </summary>
+        /// <seealso cref="HitResultExtensions.IsBasic"/>
         [Key(3)]
-        public int CountBasicHitObjects;
+        [JsonProperty("basics_count")]
+        public int BasicsCount;
```
[the remark this answers is not in the document]
Comment (frenzibyte) [reply osu.Game/Scoring/ScoringValues.cs:47]: Should the same apply to the ticks as well? I think they're less weird than this one.

[the remark this answers is not in the document]
Comment (smoogipoo) [reply osu.Game/Scoring/ScoringValues.cs:47]: Those are fine to me

**Comment by smoogipoo** [comment]
As discussed in Discord, store `base_score`/`bonus_score` directly in `ScoreInfo`/`SoloScoreInfo` instead.

**Comment by frenzibyte** [comment]
I've applied the changes as discussed. Until all scores are reprocessed to provide the newly added properties, I've kept all of them nullable and fall back to the current way of retrieving it (via diffcalc or otherwise), so that no immediate breakage is incurred from this PR. I've also added a TODO comment for `TotalScore` since we don't want this PR to get any larger in diff.

> I'm not sure about the tick stuff. Also not sure how that's going to work with stable yet. Because the goal at the end of the day is for these to never be null inside `ScoreInfo`/`SoloScoreInfo`.

I had thought lazer scores are now in a completely separate table with a different row structure from stable? Unless somewhere down the line lazer and stable scores are both processed by one central component?

File: osu.Game/Online/Spectator/FrameHeader.cs [hunk 2]
```diff
@@ -44,28 +44,43 @@ public class FrameHeader
...
         public DateTimeOffset ReceivedTime { get; set; }
 
+        [Key(5)]
+        public double? BaseScore { get; set; }
+
+        [Key(6)]
+        public double? BonusScore { get; set; }
+
+        [Key(7)]
+        public ScoringValues? Maximum { get; set; }
```
Comment (peppy) [inline osu.Game/Online/Spectator/FrameHeader.cs:54]: I think this should also be called `MaximumScoringValues`? `Maximum` on its own doesn't mean much to me.

File: osu.Game/Database/RealmAccess.cs [hunk 3]
```diff
@@ -67,8 +67,9 @@ public class RealmAccess : IDisposable
         /// 20   2022-07-21    Added LastAppliedDifficultyVersion to RulesetInfo, changed default value of BeatmapInfo.StarRating to -1.
         /// 21   2022-07-27    Migrate collections to realm (BeatmapCollection).
         /// 22   2022-07-31    Added ModPreset.
+        /// 23   2022-08-02    Added ScoringValues and MaximumScoringValues to ScoreInfo.
```
Comment (peppy) [inline osu.Game/Database/RealmAccess.cs:70]: Outdated

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Online/Spectator/FrameHeader.cs (hunk 2 above)
Comment (frenzibyte) [reply osu.Game/Online/Spectator/FrameHeader.cs:54]: Yeah I wasn't sure on the naming, felt `Maximum` means enough to signify maximum achievable value, but maybe not. Will go back to `MaximumScoringValues` for now.

**Comment by smoogipoo** [comment]
Thanks for this PR.

I've analysed the path forward for if this or other alternatives existed https://github.com/ppy/osu/pull/19939. The conclusion is that while it simplifies calculations it makes it extremely difficult to recalculate scores. It seems like the better solution is, after all, to just store the maximum statistics. The raw JSON size is also smaller doing it that way too.

**Pull request closed without merging**